installation.md: simplify build-from-source for rpm distros

lsm5 opened the pull request:
RPM distros can simply run `dnf builddep` on `rpm/podman.spec`. That
specfile already handles dependencies for all supported versions of
Fedora and CentOS Stream so it's a lot simpler and easier to maintain
than duplicating dependency specification in the documentation.

Resolves: #157

lsm5 commented:
This should go in only after https://github.com/containers/podman/pull/19214 is merged.

lsm5 commented:
hmm, `git-archive-all` would be a problem on centos and rhel. Guess I should just replace it with `git archive...`

TomSweeneyRedHat commented on docs/installation.md:
```diff
@@ -283,32 +283,14 @@ sudo dnf install podman
-  libselinux-devel \
-  make \
-  pkgconfig
+sudo dnf -y builddep rpm/podman.spec
```
Showing my builddep lack of knowledge here.  Does this really work on RHEL too?  I see notes about it in Fedora, but not in other places.  Also, the file `rpm/podman.spec` will always be in  place where dnf/builddep will know to find it?  i.e. the user won't have to get the file from somewhere.

If this all works it's great.  I just gave a quick try on my new PC and it failed with a no such file error....

lsm5 replied:
RHEL (and not CentOS Stream) could be a different beast based on `-devel` package availability. I think those need a different repo. But the `builddep` step should work everywhere dnf is present. I can add additional step to ensure `rpmbuild` and `spectool` are installed as well.

lsm5 replied:
`dnf builddep` takes a path to the rpm spec file. And these steps are run from the main podman git repo directory. rpm/podman.spec is the same file used for packit / copr /fedora tasks.

lsm5 replied:
let me try it on rhel too and update the PR. Thanks

lsm5 commented:
@TomSweeneyRedHat @Luap99 @rhatdan PTAL

@Ricky-Tigg the podman PR is merged. Please feel free to try the updated steps in this PR if you'd like.

lsm5 commented:
@cgiradkar PTAL

Luap99 commented on docs/installation.md:
```diff
@@ -283,32 +283,14 @@ sudo dnf install podman
-  libselinux-devel \
-  make \
-  pkgconfig
+sudo dnf -y builddep rpm/podman.spec
```
will this only install build deps?
What about the runtime deps such as conmon, crun, netavark, etc...? Or will the rpm-install step pull these in?

lsm5 replied:
yes, builddep only installs build dependencies. The third step `sudo make rpm-install` will do a `dnf install $RPM_PATH` which will install the runtime deps.

lsm5 replied:
I can add comments in there if it helps

TomSweeneyRedHat replied:
comments would be appreciated

lsm5 commented:
err ... looking at the title `Building from scratch` I wonder if I should revise this PR to NOT mention the rpmbuild steps but only get build and install dependencies. Sorry I just lost track.

lsm5 commented:
do we `expect` the user to want to run `make` or will the user be cool with `make rpm` ? I'm happy to go either way.

lsm5 commented:
fetching install time dependencies directly from the spec file and avoiding `rpmbuild` will likely be not so trivial, and might also end up needing frequent maintenance. Just that `dnf install $RPM` very easily handles all install dependencies.

TomSweeneyRedHat commented:
I think most folks will be looking to do 'make'.  Instead of "Building from Scratch", I'd be OK with "Building from Source" if you think that's better.

Luap99 commented:
If I just look at the current list I see several packages that are missing and/or should not be there. Thus I like the dnf builddep and dnf install podman because well we know that the rpm spec is always correct otherwise it would not build.

But then again we have the same problem on debian/ubuntu so maybe we should allocate someone to keep this in sync.

Also the normal make commands are listed under `### Get Source Code`, this section here should really only list deps to be consistent with the other distros.

lsm5 commented:
ack, I'll revise it this way then:

```
# build deps
dnf -y builddep rpm/podman.spec

# runtime deps
dnf -y install catatonit conmon containers-common-extra
```

I can leave a note in `rpm/podman.spec` saying if any dependency needs to be updated then podman.io `building from source` should be updated as well.

lsm5 commented:
So, `go-md2man` is not shipped on rhel8 / c8s. I could provide a copr to install it but that might bring in some more ugliness by polluting the user's env. 

Are we ok with limiting the RHEL steps to rhel 9 / c9s ? 

I have separated Fedora out from RHEL and CentOS Stream already as containers-common-extra is only on Fedora and RHEL/centos needs additional repos enabled.

lsm5 commented:
> So, `go-md2man` is not shipped on rhel8 / c8s. I could provide a copr to install it but that might bring in some more ugliness by polluting the user's env.
> 
> Are we ok with limiting the RHEL steps to rhel 9 / c9s ?
> 
> I have separated Fedora out from RHEL and CentOS Stream already as containers-common-extra is only on Fedora and RHEL/centos needs additional repos enabled.

@Luap99 @TomSweeneyRedHat ping, wdyt ?

Luap99 commented:
Well the current instructions (listed packages) already do not work for RHEL so this would not be a new problem.

I would love to have some basic instruction on how to build on RHEL 8 as I had to test some of my patches there in the past and it was difficult to figure out how to get the right deps. However I won't block over it. If RHEL 9 works than I see this already as good improvement.

TomSweeneyRedHat commented:
@lsm5 I'd say no copr.  I'm OK with providing instructions for RHEL9 only, but maybe we can add "try at your own risk" kind of verbiage for RHEL 8.

lsm5 commented:
ack, I'll do Fedora, RHEL 9, CentOS 9, followed by RHEL 8, CentOS 8, so 3 sections.

TomSweeneyRedHat commented:
just a quick friendly @lsm5 ping to put this back on the radar

lsm5 commented:
> just a quick friendly @lsm5 ping to put this back on the radar

oops, yup. Let me get on this one. The go-md2man issue has gone away so things are a little simpler now.

lsm5 commented:
blocking on https://github.com/containers/podman/pull/19670

lsm5 commented:
Updated PR. Have separate instructions for CentOS Stream, RHEL 8 and 9. RE: RHEL 8 and 9, I have yet to verify that on RHEL machines. The repos on rhel containers don't find the shadow-utils-subid-devel and ostree-devel packages. I'll get back to this Monday.

lsm5 commented:
> Updated PR. Have separate instructions for CentOS Stream, RHEL 8 and 9. RE: RHEL 8 and 9, I have yet to verify that on RHEL machines. The repos on rhel containers don't find the shadow-utils-subid-devel and ostree-devel packages. I'll get back to this Monday.

tried this on an internal rhel instance and the CRB and appstream repos were able to find all required packages. I'm not sure if a public RHEL instance with developer subscription or such will have access to the exact same repos, but we can cross that bridge later.

@TomSweeneyRedHat @Luap99 Good for another review.

TomSweeneyRedHat commented on docs/installation.md:
```diff
@@ -277,38 +277,46 @@ sudo dnf copr enable rhcontainerbot/podman-next -y
+
+On RHEL 8, enable the Appstream and CodeReady Builder (CRB) repositories.
+
+On CentOS Stream 9 and RHEL 9, enable the CodeReady Builder (CRB) repository.
```
It might be nice to have examples on how to ready those repos

lsm5 replied:
ack, doing it now.

TomSweeneyRedHat commented:
One small suggestion for consideration, otherwise

LGTM

lsm5 committed:
installation.md: simplify build-from-source for rpm distros

RPM distros can simply run `dnf builddep` on `rpm/podman.spec` to fetch
build dependencies. That specfile already handles dependencies
for all supported versions of Fedora and CentOS Stream so it's
a lot simpler and easier to maintain than duplicating dependency
specification in the documentation.

Resolves: #157

Signed-off-by: Lokesh Mandvekar <lsm5@fedoraproject.org>

lsm5 commented:
@TomSweeneyRedHat @Luap99 PTAL again. I simplified the `additional repo enablement` step by a lot. The `builddep` step for rhel should work with a simple copy paste.

baude commented:
lgtm

Luap99 reviewed (commented):
LGTM

TomSweeneyRedHat commented:
@lsm5 very nice rework btb, TYVM!
